Support adding Wireguard option to all applications

Mahmoud-Emad opened the pull request:
### Description

WireGuard Option Across All Applications

- Added the ability to configure WireGuard access for machines.
- Integrated the WireGuard option into all applications within the Dashboard, enhancing security and connectivity options.

### Changes

- Enable Wireguard access to the machine.
- Include the option in all Dashboard applications.

- Enabled the Public IPv4 and displayed it as a 'read-only' field
- Send the value of the 'Wireguard' option to the deployment
- Updated the return of the 'createWorker' function to be ready to accept all networks
- Updated the worker interface to accept all networks

- Infrared the type of the name in the validator to avoid typescript errors
- Send the value of the 'Wireguard' to the deployment
- Enable Mycelium option by default
- Added the Wireguard option to the K8S deployment

- Added a watcher into all solutions that require a domain to be deployed to watch on the custom domain value then updated the Wireguard option
- Added a new attr 'read-only' in the 'Networks' component to set the 'Wireguard' option as 'read-only'
- Update the opacity of the 'Wireguard Switch' in the 'Networks' component
- Updated all applications with the new requirement and tested the full workflow

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3185
- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3283
- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3282

### Screenshots

![image](https://github.com/user-attachments/assets/002cb634-ed27-4bb1-8b75-105d378894df)

### Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

Mahmoud-Emad committed:
feat: Support adding Wireguard option to all applications:

- Enable Wireguard access to the machine.
- Include the option in all Dashboard applications.

zaelgohary reviewed (changes requested):
Still missing from the following apps:

- Freeflow
- Caprover
- K8S
- Owncloud

Mahmoud-Emad committed:
feat: Add missing networks to FreeFlow solution.

Mahmoud-Emad committed:
feat: Support all networks on Caprover solution:

- Enabled the Public IPv4 and displayed it as a 'read-only' field
- Send the value of the 'Wireguard' option to the deployment
- Updated the return of the 'createWorker' function to be ready to accept all networks
- Updated the worker interface to accept all networks

Mahmoud-Emad committed:
style: Change the 'mycelium' in networks component to 'Mycelium'

Mahmoud-Emad committed:
fix: Send the 'add wireguard access' to the deployment:

- Infared the type of the name in the validator to avoid typescript errors
- Send the value of the 'wireguard' to the deployment
- Enable Mycelium option by defualt
- Added the wireguard option to the K8S deployment

Mahmoud-Emad committed:
feat: Support enabling Wireguard option when custom domain option is false:

- Added a watcher into all solutions that require a domain to be deployed to watch on the custom domain value then updated the Wireguard option
- Added a new attr 'read-only' in the 'Networks' component to set the 'Wireguard' option as 'read-only'
- Update the opacity of the 'Wireguard Switch' in the 'Networks' component
- Updated all applications with the new requirement and tested the full workflow

Mahmoud-Emad commented:
Let's use `Squash Merge` when merging this PR

0oM4R commented on packages/playground/src/weblets/tf_umbrel.vue:
i think we don't need to disable the wg here we don't have domains

0oM4R commented:
I have a concern about switching off the wg access on enabling custom domain; I mean, I checked the network section and found that the wg is enabled, then scroll down and found gateways section and enabled the custom domain... then i will figure out later that my machine doesn't have wg access what do you think. 

Also, I will open an issue for adding tooltips for each disabled option to clarify why it is readonly

0oM4R commented:
> I have a concern about switching off the wg access on enabling custom domain; I mean, I checked the network section and found that the wg is enabled, then scroll down and found gateways section and enabled the custom domain... then i will figure out later that my machine doesn't have wg access what do you think.
> 
> Also, I will open an issue for adding tooltips for each disabled option to clarify why it is readonly

I understand that the gw deployment will fail if we disable the wireguard but my point is to notify the user somehow that we disabled the wireguard

0oM4R commented:
I deployed a wordpress inctance with custom domain and switched the wg option off, but i got wg configuration on my deployment!
I think this is because the wireguard is required as long as we have gw field,
the only case we should have wg is optional is when we have ipv4 and custom domain:
![Screenshot from 2024-08-21 15-53-49](https://github.com/user-attachments/assets/5253d1a3-28f3-4f20-b152-92ec7b4fe19b)

Mahmoud-Emad committed:
Merge branch 'development' into development_wireguard_applications

Mahmoud-Emad committed:
fix: Enable wireguard if:

- Custom domain is enabled and the machine has publicIP
- Custom domain is enabled but the machine has domain on our gateways.

Mahmoud-Emad commented:
> I have a concern about switching off the wg access on enabling custom domain; I mean, I checked the network section and found that the wg is enabled, then scroll down and found gateways section and enabled the custom domain... then i will figure out later that my machine doesn't have wg access what do you think.
> 
> Also, I will open an issue for adding tooltips for each disabled option to clarify why it is readonly

please do

0oM4R commented:
> I have a concern about switching off the wg access on enabling custom domain; I mean, I checked the network section and found that the wg is enabled, then scroll down and found gateways section and enabled the custom domain... then i will figure out later that my machine doesn't have wg access what do you think.

both parts ?

0oM4R commented:
> > I have a concern about switching off the wg access on enabling custom domain; I mean, I checked the network section and found that the wg is enabled, then scroll down and found gateways section and enabled the custom domain... then i will figure out later that my machine doesn't have wg access what do you think.
> > Also, I will open an issue for adding tooltips for each disabled option to clarify why it is readonly
> 
> please do




 @Mahmoud-Emad  both parts ?

Mahmoud-Emad committed:
docs: Added a hint message on the wreguard network option

amiraabouhadid commented on packages/playground/src/components/networks.vue:
```diff
@@ -68,22 +68,24 @@
               v-if="wireguard !== null"
               inline
-              tooltip="Enabling WireGuard Access allows you to establish private, secure, and encrypted connections to your instance."
+              tooltip="Enabling WireGuard Access allows you to establish private, secure, and encrypted connections to your instance. Please note that Unless you use the custom domain with an IPV4, this field will be read-only."
```
should be 
```suggestion
              tooltip="Enabling WireGuard Access allows you to establish private, secure, and encrypted instance connections. Please note that this field will be read-only unless you use a custom domain with IPV4."
```

amiraabouhadid commented on packages/playground/src/weblets/full_vm.vue:
```diff
@@ -25,7 +25,7 @@
             validators.required('Name is required.'),
             validators.IsAlphanumericExpectUnderscore('Name should consist of letters ,numbers and underscores only.'),
-            name => validators.isAlpha('Name must start with alphabet char.')(name[0]),
+            (name: string) => validators.isAlpha('Name must start with alphabet char.')(name[0]),
```
```suggestion
            (name: string) => validators.isAlpha('Name must start with an alphabetical character.')(name[0]),
```

amiraabouhadid commented on packages/playground/src/weblets/full_vm.vue:
```diff
@@ -102,7 +102,7 @@
             :rules="[
               validators.required('Disk name is required.'),
-              name => validators.isAlpha('Name must start with alphabet char.')(name[0]),
+              (name: string) => validators.isAlpha('Name must start with alphabet char.')(name[0]),
```
```suggestion
            (name: string) => validators.isAlpha('Name must start with an alphabetical character.')(name[0]),
```

amiraabouhadid commented on packages/playground/src/weblets/full_vm.vue:
```diff
@@ -102,7 +102,7 @@
               validators.required('Disk name is required.'),
-              name => validators.isAlpha('Name must start with alphabet char.')(name[0]),
+              (name: string) => validators.isAlpha('Name must start with alphabet char.')(name[0]),
               validators.minLength('Disk minLength is 2 chars.', 2),
```
```suggestion
              validators.minLength('Disk name minimum length is 2 characters.', 2),
```

amiraabouhadid commented on packages/playground/src/weblets/full_vm.vue:
```diff
@@ -102,7 +102,7 @@
-              name => validators.isAlpha('Name must start with alphabet char.')(name[0]),
+              (name: string) => validators.isAlpha('Name must start with alphabet char.')(name[0]),
               validators.minLength('Disk minLength is 2 chars.', 2),
               validators.isAlphanumeric('Disk name only accepts alphanumeric chars.'),
```
```suggestion
              validators.isAlphanumeric('Disk name only accepts alphanumeric characters.'),
```

amiraabouhadid commented on packages/playground/src/weblets/full_vm.vue:
```diff
@@ -102,7 +102,7 @@
+              (name: string) => validators.isAlpha('Name must start with alphabet char.')(name[0]),
               validators.minLength('Disk minLength is 2 chars.', 2),
               validators.isAlphanumeric('Disk name only accepts alphanumeric chars.'),
               validators.maxLength('Disk maxLength is 50 chars.', 50),
```
```suggestion
              validators.maxLength('Disk name maximum length is 50 characters.', 50),
```

amiraabouhadid commented on packages/playground/src/weblets/micro_vm.vue:
```diff
@@ -27,7 +27,7 @@
             validators.required('Name is required.'),
             validators.IsAlphanumericExpectUnderscore('Name should consist of letters ,numbers and underscores only.'),
-            name => validators.isAlpha('Name must start with alphabet char.')(name[0]),
+            (name: string) => validators.isAlpha('Name must start with alphabet char.')(name[0]),
```
```suggestion
            (name: string) => validators.isAlpha('Name must start with alphabetical character.')(name[0]),
```

amiraabouhadid commented on packages/playground/src/weblets/micro_vm.vue:
```diff
@@ -91,7 +91,7 @@
             :rules="[
               validators.required('Key name is required.'),
-              key => validators.isAlpha('Key must start with alphabet char.')(key[0]),
+              (key: string) => validators.isAlpha('Key must start with alphabet char.')(key[0]),
```
```suggestion
              (key: string) => validators.isAlpha('Key must start with alphabetical character.')(key[0]),
```



please fix all other similar validation messages accordingly

amiraabouhadid commented on packages/playground/src/weblets/micro_vm.vue:
```diff
@@ -91,7 +91,7 @@
-              key => validators.isAlpha('Key must start with alphabet char.')(key[0]),
+              (key: string) => validators.isAlpha('Key must start with alphabet char.')(key[0]),
               validators.pattern('Invalid key format.', { pattern: /^[^0-9_\s][a-zA-Z0-9_]+$/ }),
               validators.maxLength('Key max length is 128 chars.', 128),
```
```suggestion
              validators.maxLength('Key maximum length is 128 characters.', 128),
```

Mahmoud-Emad committed:
Merge branch 'development' into development_wireguard_applications

Mahmoud-Emad committed:
chore: Apply comments:

- Updated the validation messages.

0oM4R commented on packages/playground/src/weblets/tf_casperlabs.vue:
```diff
@@ -36,7 +36,14 @@
+        v-model:mycelium="mycelium"
+        v-model:ipv6="ipv6"
+        v-model:wireguard="wireguard"
+        :readOnlyWireGuard="!selectionDetails?.domain?.enabledCustomDomain"
```
we should add IPv4 to this condition

0oM4R commented on packages/playground/src/weblets/freeflow.vue:
```diff
@@ -210,10 +223,22 @@ async function deploy() {
 }
+
+watch(
+  () => selectionDetails.value?.domain?.enabledCustomDomain,
```
can we move this logic to the network tab?
instead of passing readonly flag, i suggest to pass the customDomainEnabled and check for both of customdomain enabled and ipv4 values then handle the wireguard option

Mahmoud-Emad replied:
I have an open issue for enhancing the entire component, please if you want to do this enhancement in this PR, check with @AhmedHanafy725

https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3294

Mahmoud-Emad replied:
It's also a valid comment tbh, we can do it here to avoid code redundancy

Mahmoud-Emad replied:
- Established a new constant by computing the options for IPv4 and custom domains to decide whether the Wireguard option should be read-only or not.
- Monitored the 'isWireguardReadonly' constant to determine the value of the Wireguard feature based on it.

Mahmoud-Emad committed:
fix: Move the logic of enabling/disabling the Wireguard option to the network component

Mahmoud-Emad committed:
fix: Move the Wireguard watcher into the networks component.

Mahmoud-Emad committed:
feat: Enhance the enabling Wireguard functionality:

- Established a new constant by computing the options for IPv4 and custom domains to decide whether the Wireguard option should be read-only or not.
- Monitored the 'isWireguardReadonly' constant to determine the value of the Wireguard feature based on it.

0oM4R commented on packages/playground/src/weblets/tf_presearch.vue:
i found the wg option is disabled while we don't have ipv4 here, but the idea is we are not using the gateway

Mahmoud-Emad committed:
fix: Enhance the network component

- Set the wireguard option as read-only in case of the solution doesn't require domain.
- Send the full domain object to the network component

Mahmoud-Emad committed:
chore: Fix build
